# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1417532210 18000
#      Tue Dec 02 09:56:50 2014 -0500
# Branch gui-release
# Node ID 264ff6bf74753571af7d4eeda39ce4d30779fe01
# Parent  27f5d1607b8484d08b3759aa7ada6d8d06fcf1dd
use gnulib:: namespace for log, logf, localtime, and gmtime

* load-save.cc, lo-mappers.cc, lo-specfun.cc, oct-time.cc: Sprinkle
with gnulib:: to use log, logf, localtime, and gmtime from gnulib if
they are needed.

diff --git a/libinterp/corefcn/load-save.cc b/libinterp/corefcn/load-save.cc
--- a/libinterp/corefcn/load-save.cc
+++ b/libinterp/corefcn/load-save.cc
@@ -1214,17 +1214,17 @@ write_header (std::ostream& os, load_sav
       {
         char const * versionmagic;
         int16_t number = *(reinterpret_cast<const int16_t *>("\x00\x01"));
         struct tm bdt;
         time_t now;
         char headertext[128];
 
         time (&now);
-        bdt = *gmtime (&now);
+        bdt = *gnulib::gmtime (&now);
         memset (headertext, ' ', 124);
         // ISO 8601 format date
         nstrftime (headertext, 124, "MATLAB 5.0 MAT-file, written by Octave "
                    OCTAVE_VERSION ", %Y-%m-%d %T UTC", &bdt, 1, 0);
 
         // The first pair of bytes give the version of the MAT file
         // format.  The second pair of bytes form a magic number which
         // signals a MAT file.  MAT file data are always written in
diff --git a/liboctave/numeric/lo-mappers.cc b/liboctave/numeric/lo-mappers.cc
--- a/liboctave/numeric/lo-mappers.cc
+++ b/liboctave/numeric/lo-mappers.cc
@@ -517,24 +517,26 @@ rc_atanh (float x)
   return fabsf (x) > 1.0f ? atanh (FloatComplex (x))
                           : FloatComplex (atanhf (x));
 }
 
 Complex
 rc_log (double x)
 {
   const double pi = 3.14159265358979323846;
-  return x < 0.0 ? Complex (log (-x), pi) : Complex (log (x));
+  return x < 0.0 ? Complex (gnulib::log (-x), pi) : Complex (gnulib::log (x));
 }
 
 FloatComplex
 rc_log (float x)
 {
   const float pi = 3.14159265358979323846f;
-  return x < 0.0f ? FloatComplex (logf (-x), pi) : FloatComplex (logf (x));
+  return (x < 0.0f
+          ? FloatComplex (gnulib::logf (-x), pi)
+          : FloatComplex (gnulib::logf (x)));
 }
 
 Complex
 rc_log2 (double x)
 {
   const double pil2 = 4.53236014182719380962; // = pi / log(2)
   return x < 0.0 ? Complex (xlog2 (-x), pil2) : Complex (xlog2 (x));
 }
diff --git a/liboctave/numeric/lo-specfun.cc b/liboctave/numeric/lo-specfun.cc
--- a/liboctave/numeric/lo-specfun.cc
+++ b/liboctave/numeric/lo-specfun.cc
@@ -628,17 +628,17 @@ log1p (double x)
       // approximation log (1+x) ~ 2*sum ((x/(2+x)).^ii ./ ii), ii = 1:2:2n+1
       double u = x / (2 + x), t = 1, s = 0;
       for (int i = 2; i < 12; i += 2)
         s += (t *= u*u) / (i+1);
 
       retval = 2 * (s + 1) * u;
     }
   else
-    retval = log (1 + x);
+    retval = gnulib::log (1 + x);
 
   return retval;
 }
 #endif
 
 Complex
 log1p (const Complex& x)
 {
@@ -680,24 +680,24 @@ log1pf (float x)
 {
   float retval;
 
   float ax = fabs (x);
 
   if (ax < 0.2)
     {
       // approximation log (1+x) ~ 2*sum ((x/(2+x)).^ii ./ ii), ii = 1:2:2n+1
-      float u = x / (2 + x), t = 1, s = 0;
+      float u = x / (2 + x), t = 1.0f, s = 0;
       for (int i = 2; i < 12; i += 2)
         s += (t *= u*u) / (i+1);
 
-      retval = 2 * (s + 1) * u;
+      retval = 2 * (s + 1.0f) * u;
     }
   else
-    retval = log (1 + x);
+    retval = gnulib::logf (1.0f + x);
 
   return retval;
 }
 #endif
 
 FloatComplex
 log1p (const FloatComplex& x)
 {
@@ -2981,24 +2981,27 @@ done:
 
   return retval;
 }
 
 
 Complex rc_log1p (double x)
 {
   const double pi = 3.14159265358979323846;
-  return x < -1.0 ? Complex (log (-(1.0 + x)), pi) : Complex (log1p (x));
+  return (x < -1.0
+          ? Complex (gnulib::log (-(1.0 + x)), pi)
+          : Complex (log1p (x)));
 }
 
 FloatComplex rc_log1p (float x)
 {
   const float pi = 3.14159265358979323846f;
-  return x < -1.0f ? FloatComplex (logf (-(1.0f + x)), pi)
-                   : FloatComplex (log1pf (x));
+  return (x < -1.0f
+          ? FloatComplex (gnulib::logf (-(1.0f + x)), pi)
+          : FloatComplex (log1pf (x)));
 }
 
 // This algorithm is due to P. J. Acklam.
 // See http://home.online.no/~pjacklam/notes/invnorm/
 // The rational approximation has relative accuracy 1.15e-9 in the whole region.
 // For doubles, it is refined by a single step of Halley's 3rd order method.
 // For single precision, the accuracy is already OK, so we skip it to get
 // faster evaluation.
@@ -3041,17 +3044,17 @@ static double do_erfinv (double x, bool 
       const double q = 0.5 * x, r = q*q;
       const double yn = (((((a[0]*r + a[1])*r + a[2])*r + a[3])*r + a[4])*r + a[5])*q;
       const double yd = ((((b[0]*r + b[1])*r + b[2])*r + b[3])*r + b[4])*r + 1.0;
       y = yn / yd;
     }
   else if (ax < 1.0)
     {
       // Tail region.
-      const double q = sqrt (-2*log (0.5*(1-ax)));
+      const double q = sqrt (-2*gnulib::log (0.5*(1-ax)));
       const double yn = ((((c[0]*q + c[1])*q + c[2])*q + c[3])*q + c[4])*q + c[5];
       const double yd = (((d[0]*q + d[1])*q + d[2])*q + d[3])*q + 1.0;
       y = yn / yd * signum (-x);
     }
   else if (ax == 1.0)
     return octave_Inf * signum (x);
   else
     return octave_NaN;
@@ -3119,20 +3122,26 @@ static double do_erfcinv (double x, bool
       const double q = 0.5*(1-x), r = q*q;
       const double yn = (((((a[0]*r + a[1])*r + a[2])*r + a[3])*r + a[4])*r + a[5])*q;
       const double yd = ((((b[0]*r + b[1])*r + b[2])*r + b[3])*r + b[4])*r + 1.0;
       y = yn / yd;
     }
   else if (x > 0.0 && x < 2.0)
     {
       // Tail region.
-      const double q = x < 1 ? sqrt (-2*log (0.5*x)) : sqrt (-2*log (0.5*(2-x)));
+      const double q = (x < 1
+                        ? sqrt (-2*gnulib::log (0.5*x))
+                        : sqrt (-2*gnulib::log (0.5*(2-x))));
+
       const double yn = ((((c[0]*q + c[1])*q + c[2])*q + c[3])*q + c[4])*q + c[5];
+
       const double yd = (((d[0]*q + d[1])*q + d[2])*q + d[3])*q + 1.0;
+
       y = yn / yd;
+
       if (x < pbreak_lo)
         y = -y;
     }
   else if (x == 0.0)
     return octave_Inf;
   else if (x == 2.0)
     return -octave_Inf;
   else
@@ -3238,18 +3247,18 @@ betain (double x, double p, double q, do
   for ( ; ; )
     {
       term = term * temp * rx / (pp + ai);
       value = value + term;
       temp = fabs (term);
 
       if (temp <= acu && temp <= acu * value)
         {
-          value = value * exp (pp * log (xx)
-                               + (qq - 1.0) * log (cx) - beta) / pp;
+          value = value * exp (pp * gnulib::log (xx)
+                               + (qq - 1.0) * gnulib::log (cx) - beta) / pp;
 
           if (indx)
             {
               value = 1.0 - value;
             }
           break;
         }
 
@@ -3337,17 +3346,17 @@ betaincinv (double y, double p, double q
       a = y;
       pp = p;
       qq = q;
       indx = false;
     }
 
   //  Calculate the initial approximation.
 
-  r = sqrt (- log (a * a));
+  r = sqrt (- gnulib::log (a * a));
 
   ycur = r - (2.30753 + 0.27061 * r) / (1.0 + (0.99229 + 0.04481 * r) * r);
 
   if (1.0 < pp && 1.0 < qq)
     {
       r = (ycur * ycur - 3.0) / 6.0;
       s = 1.0 / (pp + pp - 1.0);
       t = 1.0 / (qq + qq - 1.0);
@@ -3358,25 +3367,25 @@ betaincinv (double y, double p, double q
   else
     {
       r = qq + qq;
       t = 1.0 / (9.0 * qq);
       t = r * pow (1.0 - t + ycur * sqrt (t), 3);
 
       if (t <= 0.0)
         {
-          value = 1.0 - exp ((log ((1.0 - a) * qq) + beta) / qq);
+          value = 1.0 - exp ((gnulib::log ((1.0 - a) * qq) + beta) / qq);
         }
       else
         {
           t = (4.0 * pp + r - 2.0) / t;
 
           if (t <= 1.0)
             {
-              value = exp ((log (a * pp) + beta) / pp);
+              value = exp ((gnulib::log (a * pp) + beta) / pp);
             }
           else
             {
               value = 1.0 - 2.0 / (t + 1.0);
             }
         }
     }
 
@@ -3408,17 +3417,17 @@ betaincinv (double y, double p, double q
       ycur = betain (value, pp, qq, beta, err);
 
       if (err)
         {
           return value;
         }
 
       xin = value;
-      ycur = (ycur - a) * exp (beta + r * log (xin) + t * log (1.0 - xin));
+      ycur = (ycur - a) * exp (beta + r * gnulib::log (xin) + t * gnulib::log (1.0 - xin));
 
       if (ycur * yprev <= 0.0)
         {
           prev = std::max (sq, fpu);
         }
 
       g = 1.0;
 
diff --git a/liboctave/system/oct-time.cc b/liboctave/system/oct-time.cc
--- a/liboctave/system/oct-time.cc
+++ b/liboctave/system/oct-time.cc
@@ -230,27 +230,27 @@ octave_base_tm::init (void *p)
 
 void
 octave_localtime::init (const octave_time& ot)
 {
   tm_usec = ot.usec ();
 
   time_t t = ot.unix_time ();
 
-  octave_base_tm::init (localtime (&t));
+  octave_base_tm::init (gnulib::localtime (&t));
 }
 
 void
 octave_gmtime::init (const octave_time& ot)
 {
   tm_usec = ot.usec ();
 
   time_t t = ot.unix_time ();
 
-  octave_base_tm::init (gmtime (&t));
+  octave_base_tm::init (gnulib::gmtime (&t));
 }
 
 void
 octave_strptime::init (const std::string& str, const std::string& fmt)
 {
   struct tm t;
 
   t.tm_sec = 0;
